concent_api/core/models.py

```diff
@@ -73,3 +76,167 @@ class MessageAuth(Model):
+        ],
+        SubtaskState.FAILED.name: [  # pylint: disable=no-member
+            SubtaskState.FORCING_REPORT.name,  # pylint: disable=no-member
+            SubtaskState.FORCING_RESULT_TRANSFER.name,  # pylint: disable=no-member
+            SubtaskState.ADDITIONAL_VERIFICATION.name,  # pylint: disable=no-member
+        ],
+    }
+
+    class Meta:
+        unique_together = (
+            ('requestor', 'task_id'),
+            ('requestor', 'subtask_id'),
+        )
+
+    task_id                     = models.CharField(max_length = MESSAGE_TASK_ID_MAX_LENGTH)
+    subtask_id                  = models.CharField(max_length = MESSAGE_TASK_ID_MAX_LENGTH, unique = True, db_index = True)
+
+    # Relation to concent client in requestor context.
+    requestor                   = models.ForeignKey(Client, related_name = 'subtasks_as_requestor')
+
+    # Relation to concent client in provider context.
+    provider                    = models.ForeignKey(Client, related_name = 'subtasks_as_provider')
+
+    state                       = models.CharField(max_length = 32, choices = SubtaskState.choices())
+    _current_state              = None
+
+    # If in an active state, it's the time at which Concent automatically transitions to a passive state if it does
+    # not get the information it expects from the client (a timeout). Must be NULL in a passive state.
+    next_deadline               = models.DateTimeField(blank = True, null = True)
+
+    # Related messages
+    task_to_compute             = models.OneToOneField(StoredMessage, blank = True, null = True, related_name = 'subtasks_for_task_to_compute')
+    report_computed_task        = models.OneToOneField(StoredMessage, blank = True, null = True, related_name = 'subtasks_for_report_computed_task')
+    ack_report_computed_task    = models.OneToOneField(StoredMessage, blank = True, null = True, related_name = 'subtasks_for_ack_report_computed_task')
+    reject_report_computed_task = models.OneToOneField(StoredMessage, blank = True, null = True, related_name = 'subtasks_for_reject_report_computed_task')
+    subtask_results_accepted    = models.OneToOneField(StoredMessage, blank = True, null = True, related_name = 'subtasks_for_subtask_results_accepted')
+    subtask_results_rejected    = models.OneToOneField(StoredMessage, blank = True, null = True, related_name = 'subtasks_for_subtask_results_rejected')
+
+    def __init__(self, *args, **kwargs):
+        super(Subtask, self).__init__(*args, **kwargs)
+        if self.pk is not None:
+            self._current_state = self.state
+
+    def clean(self):
+        super(Subtask, self).clean()
+
+        # next_deadline can be None only for passive states
+        if self.pk is not None and self.next_deadline is None and self.state not in self.SUBTASK_PASSIVE_STATES:
+            raise ValidationError({
+                'next_deadline': 'next_deadline is None for active state.'
+            })
```

Comment:
:+1:

Reply:
:+1: 

> States in which task_to_compute, report_computed_task and other foreign keys to messages may and may not be None. It might be tedious to define but I think it's likely to catch lots of invalid transitions - where you change state but forget to store the required messages in the table.

Added, but please review it when committed.

Reply:
> Validate that task_compute is the same as the one inside report_computed_task if the latter is not None. Same for any other nested messages.

I suggest creating a separate issue, because it sounds like lot of work and thinking through.

Reply:
Issue created: https://github.com/golemfactory/concent/issues/177

Not in pivotal yet because i think it should be grouped in a single task with other left-overs from this one.